opwen_email_server/mailers/shared.py

```diff
@@ -0,0 +1,47 @@
+from typing import Callable
+from typing import Dict
+
+from libcloud.storage.types import ObjectDoesNotExistError
+
+from opwen_email_server.actions import _Action
+from opwen_email_server.services.storage import AzureObjectStorage
+from opwen_email_server.services.storage import AzureTextStorage
+from opwen_email_server.utils.email_parser import MimeEmailParser
+from opwen_email_server.utils.unique import new_email_id
+
+
+class ProcessServiceEmail(_Action):
+    def __init__(self,
+                 raw_email_storage: AzureTextStorage,
+                 email_storage: AzureObjectStorage,
+                 next_task: Callable[[str], None],
+                 registry: Dict[str, Callable[[dict], dict]],
+                 email_parser: Callable[[dict], dict] = None):
+
+        self._raw_email_storage = raw_email_storage
+        self._email_storage = email_storage
+        self._next_task = next_task
+        self._registry = registry
+        self._email_parser = email_parser or MimeEmailParser()
+
+    def _action(self, resource_id):  # type: ignore
+        try:
+            mime_email = self._raw_email_storage.fetch_text(resource_id)
+        except ObjectDoesNotExistError:
+            self.log_warning('Inbound email %s does not exist', resource_id)
+            return 'skipped', 202
+
+        email = self._email_parser(mime_email)
+
+        for address in email['to']:
+            mailer_service = self._registry[address]
+            formatted_email = mailer_service(email)
```

Comment:
What happens if a `to` address doesn't exist in the registry? I reckon we should skip and log any unknown addresses in case someone made a typo and/or made a mistake in including multiple recipients in the email. We could do something like this:

```suggestion
            try:
                mailer_service = self._registry[address]
            except KeyError:
                self.log_warning('Skipping unknown mailer service: %s', address)
                continue

            formatted_email = mailer_service(email)
```

Note the use of `try/except` instead of `if key in ...` due to the [EAFP vs LBYL principle](https://devblogs.microsoft.com/python/idiomatic-python-eafp-versus-lbyl/).

Reply:
Okay! Oh interesting, thanks for the insight!